services/placement/algo/mirrored.go

```diff
@@ -0,0 +1,299 @@
+	}
+
+	shards := leavingInstance.Shards()
+	for _, s := range shards.All() {
+		// NB(cw): Mark Available as Initializing in case we use shard state
+		// to indicate cutover/cutoff time for the shard.
+		if s.State() == shard.Available {
+			s.SetState(shard.Initializing)
+		}
+	}
+
+	addingInstance = addingInstance.SetShards(shards).SetShardSetID(leavingInstance.ShardSetID())
+
+	// NB(cw): Replace is supposed to be used when the leaving node is considered dead,
+	// so the leaving node will be removed from the placement in this algo, unlike the
+	// replacement of non-mirrored algo which will leave the leaving node in the
+	// placement with all shards marked as Leaving.
+	// This is because we use shardSetID as the sourceID in mirrored placement for
+	// initializing shards, and I don't want the adding instance to keep a sourceID
+	// of the leaving instance for this case.
+	instances := placement.RemoveInstanceFromList(p.Instances(), instanceID)
+	instances = append(instances, addingInstance)
+	return placement.NewPlacement().
+		SetInstances(instances).
+		SetReplicaFactor(p.ReplicaFactor()).
+		SetShards(p.Shards()).
+		SetIsSharded(p.IsSharded()).
+		SetIsMirrored(p.IsMirrored()), nil
+}
+
+func groupInstancesByShardSetID(
+	instances []services.PlacementInstance,
+	rf int,
+) ([]services.PlacementInstance, error) {
+	var (
+		shardSetMap   = make(map[string]uint32, len(instances))
+		shardSetCount = make(map[string]int, len(instances))
+		res           = make([]services.PlacementInstance, 0, len(instances))
+	)
+	for _, instance := range instances {
+		ssID := instance.ShardSetID()
+		weight := instance.Weight()
+		shardSetCount[ssID] = shardSetCount[ssID] + 1
+		if w, ok := shardSetMap[ssID]; ok {
+			if weight == w {
+				continue
+			}
+			return nil, fmt.Errorf("found different weights: %d and %d, for shardset id %s", w, weight, ssID)
+		}
+
+		shardSetMap[ssID] = instance.Weight()
+		res = append(
+			res,
+			placement.NewInstance().
+				SetID(ssID).
+				SetRack(ssID).
+				SetWeight(weight).
+				SetShardSetID(instance.ShardSetID()).
+				SetShards(instance.Shards()),
+		)
+	}
+
+	for ssID, count := range shardSetCount {
+		if count != rf {
+			return nil, fmt.Errorf("got %d count of shard set id %s, expecting %d", count, ssID, rf)
+		}
+	}
+
+	return res, nil
+}
+
+// mirrorFromPlacement zips all instances with the same shardSetID into a virtual instance
+// and create a placement with those virtual instance and rf=1.
+func mirrorFromPlacement(p services.Placement) (services.Placement, error) {
+	mirrorInstances, err := groupInstancesByShardSetID(p.Instances(), p.ReplicaFactor())
+	if err != nil {
+		return nil, err
+	}
+
+	mirror := placement.NewPlacement().
+		SetInstances(mirrorInstances).
+		SetReplicaFactor(1).
+		SetShards(p.Shards()).
+		SetIsSharded(true).
+		SetIsMirrored(true)
+	// NB(cw): Always clean up the shard states from previous placement changes,
```

Comment:
that's correct, what I meant is, for example, when the replace just happened, the RF for the shardset became 3 at that point, but the placement is supposed to have a shardset of 2, now you realize the new instance is having problem and could not take shards over/mark them as available. Now you want to revert the replace, so you call the replace function reversely again, if you look at the current algorithm, in the beginning of each operation, it has to generate a mirror placement from the original placement, in this case, it would report an error when trying to generate the mirror for the shardset because it has 3 instances for the shardset while the algorithm is expecting only 2, also the 3 instances have different status in the shards, so I have to clean up the shard states from previous placement changes before applying the next change.